### DISCUSSION TITLE
how to use the simulator to inject error

### DISCUSSION THREAD

**Discussion by lyc200150** [madsim-rs/madsim#157] [Q&A] [answered]
Hi, I'm using madsim 0.2.22，I am trying to inject some error to my project, (like some network error), but I don't know how to do. I already replace tokio with madsim-tokio，how can I inject some error then? thanks for your help!

**Comment by wangrunji0408** [comment, marked as answer]
Hi. To use the simulator, you first need to setup a virtual cluster in your test, for which you can refer to [tonic-example](https://github.com/madsim-rs/madsim/blob/main/tonic-example/tests/test.rs). After that you can call APIs in [`NetSim`](https://docs.rs/madsim/0.2.22/madsim/net/struct.NetSim.html) to inject errors, including blocking links, setting latencies etc.
